File: 2020/Java/src/aoc/Day16.java
package aoc;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Predicate;

public class Day16 {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists() || input.isDirectory()) {
			System.out.println(input.getPath() + " does not exist!");
			input = new File(input.getPath() + ".txt");
			System.out.println("trying " + input.getPath() + " instead.");
		}

		if (!input.exists() || input.isDirectory()) {
			System.out.println(input.getPath() + " does not exist!");
			input = new File(input.getParentFile().getParent(), "input");
			input = new File(input, "Day16.txt");
			System.out.println("trying " + input.getPath() + " instead.");
		}

		if (!input.exists() || input.isDirectory()) {
			System.err.println("None of the expected inputs exist!");
			return;
		} else {
			System.out.println("Using input file " + input.getPath() + ".");
		}

		int[] ownTicket = null;
		Set<int[]> tickets = new HashSet<>();
		Map<String, Predicate<Integer>> inputValidators = new HashMap<>();
		boolean your = false;
		for (String line : Files.readAllLines(input.toPath())) {
			if (line.isEmpty())
				continue;

			if (line.startsWith("your")) {
				your = true;
				continue;
			} else if (line.startsWith("nearby")) {
				your = false;
				continue;
			}

			if (!your && ownTicket == null) {
				int[] range = Arrays.stream(line.substring(line.indexOf(':') + 2).split("-| or "))
						.mapToInt(Integer::parseInt).toArray();
				inputValidators.put(line.substring(0, line.indexOf(':')),
						in -> (in >= range[0] && in <= range[1]) || (in >= range[2] && in <= range[3]));
			} else if (your) {
				ownTicket = Arrays.stream(line.split(",")).mapToInt(Integer::parseInt).toArray();
			} else {
				tickets.add(Arrays.stream(line.split(",")).mapToInt(Integer::parseInt).toArray());
			}
		}

		int invalid = 0;
		HashSet<int[]> invalidTickets = new HashSet<int[]>();
		for (int[] ticket : tickets) {
			for (int field : ticket) {
				if (inputValidators.values().stream().noneMatch(validator -> validator.test(field))) {
					invalid += field;
					invalidTickets.add(ticket);
				}
			}
		}

		System.out.println("Found an error rate of " + invalid + ".");

		for (int[] ticket : invalidTickets) {
			tickets.remove(ticket);
		}

		Map<String, Set<Integer>> possiblePositions = new HashMap<>();
		for (Entry<String, Predicate<Integer>> field : inputValidators.entrySet()) {
			for (int i = 0; i < ownTicket.length; i++) {
				final int pos = i;
				if (!field.getValue().test(ownTicket[i]))
					continue;
				else if (!tickets.stream().allMatch(t -> field.getValue().test(t[pos])))
					continue;
				else {
					Set<Integer> positions = new HashSet<>();
					if (possiblePositions.containsKey(field.getKey()))
						positions = possiblePositions.get(field.getKey());

					positions.add(i);
					possiblePositions.put(field.getKey(), positions);
				}
			}
		}

		String[] fields = new String[ownTicket.length];
		while (Arrays.stream(fields).anyMatch(f -> f == null)) {
			for (int i = 0; i < ownTicket.length; i++) {
				final int pos = i;
				if (possiblePositions.values().stream().filter(v -> v.contains(pos)).count() == 1) {
					fields[i] = possiblePositions.keySet().stream().filter(f -> possiblePositions.get(f).contains(pos))
							.findAny().get();
					possiblePositions.forEach((k, v) -> v.remove(pos));
				}
			}

			possiblePositions.keySet().stream().filter(f -> possiblePositions.get(f).size() == 1).forEach(field -> {
				final int pos = possiblePositions.get(field).iterator().next();
				fields[pos] = field;
				possiblePositions.forEach((k, v) -> v.remove(pos));
			});
		}

		long result = 1;
		for (int i = 0; i < ownTicket.length; i++) {
			if (fields[i].startsWith("departure")) {
				result *= ownTicket[i];
			}
		}

		System.out.println("The part 2 result is " + result + ".");
	}
}

File: 2020/Python/Day16.py
import os
import sys
import re


def main():
    inf = os.path.dirname(os.getcwd())
    inf = os.path.join(inf, "input")
    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf = os.path.join(os.path.dirname(inf), "input")
        inf = os.path.join(inf, "Day16.txt")
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        return
    else:
        print(f"Using input file {inf}.")

    inputFile = open(inf, 'r')

    ownTicket = []
    tickets = []
    inputValidators = {}
    validator = lambda i, range: (i >= range[0] and i <= range[1]) or (i >= range[2] and i <= range[3])
    your = False
    for line in inputFile.readlines():
        if len(line) < 3:
            continue

        if line.startswith("your"):
            your = True
            continue
        elif line.startswith("nearby"):
            your = False
            continue

        if not your and len(ownTicket) == 0:
            inputRange = [int(val) for val in re.split("-| or ", line[line.index(":") + 2:-1])]
            inputValidators[line[:line.index(":")]] = inputRange
        elif your:
            ownTicket = [int(field) for field in line[:-1].split(",")]
        else:
            ticket = [int(field) for field in line[:-1].split(",")]
            tickets.append(ticket)

    inputFile.close()

    invalid = 0
    invalidTickets = []
    for ticket in tickets:
        for field in ticket:
            if len([0 for inputRange in inputValidators.values() if validator(field, inputRange)]) == 0:
                invalid += field
                invalidTickets.append(ticket)

    print(f"Found an error rate of {invalid}.")

    for ticket in invalidTickets:
        tickets.remove(ticket)

    possiblePositions = {}
    for field in inputValidators:
        for i in range(len(ownTicket)):
            if not validator(ownTicket[i], inputValidators[field]):
                continue
            elif not all(validator(ticket[i], inputValidators[field]) for ticket in tickets):
                continue
            else:
                positions = set()
                if field in possiblePositions:
                    positions = possiblePositions[field]

                positions.add(i)
                possiblePositions[field] = positions

    fields = [None] * len(ownTicket)
    while None in fields:
        for i in range(len(ownTicket)):
            if [i in v for v in possiblePositions.values()].count(True) == 1:
                fields[i] = next(field for field in possiblePositions if i in possiblePositions[field])
                for field in possiblePositions:
                    if i in possiblePositions[field]:
                        possiblePositions[field].remove(i)

        for field in [field for field in possiblePositions if len(possiblePositions[field]) == 1]:
            if len(possiblePositions[field]) == 0:
                continue

            pos = next(iter(possiblePositions[field]))
            fields[pos] = field
            for field in possiblePositions:
                if pos in possiblePositions[field]:
                    possiblePositions[field].remove(pos)

    result = 1
    for i in range(len(ownTicket)):
        if fields[i].startswith("departure"):
            result *= ownTicket[i]

    print(f"The part 2 result is {result}.")


if __name__ == '__main__':
    main()
